src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameNodeDefinitions.cs

```diff
@@ -0,0 +1,426 @@
+
+        public StackFrameArrayTypeExpression(StackFrameNameNode typeIdentifier, ImmutableArray<StackFrameArrayRankSpecifier> arrayExpressions) : base(StackFrameKind.ArrayTypeExpression)
+        {
+            Contract.ThrowIfTrue(arrayExpressions.IsDefaultOrEmpty);
+            TypeIdentifier = typeIdentifier;
+            ArrayExpressions = arrayExpressions;
+        }
+
+        internal override int ChildCount => 1 + ArrayExpressions.Length;
+
+        public override void Accept(IStackFrameNodeVisitor visitor)
+            => visitor.Visit(this);
+
+        internal override StackFrameNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => TypeIdentifier,
+                _ => ArrayExpressions[index - 1]
+            };
+    }
+
+    internal sealed class StackFrameArrayRankSpecifier : StackFrameNode
+    {
+        public readonly StackFrameToken OpenBracket;
+        public readonly StackFrameToken CloseBracket;
+        public readonly ImmutableArray<StackFrameToken> CommaTokens;
+
+        public StackFrameArrayRankSpecifier(StackFrameToken openBracket, StackFrameToken closeBracket, ImmutableArray<StackFrameToken> commaTokens)
+            : base(StackFrameKind.ArrayExpression)
+        {
+            Contract.ThrowIfTrue(commaTokens.IsDefault);
+            Debug.Assert(openBracket.Kind == StackFrameKind.OpenBracketToken);
+            Debug.Assert(closeBracket.Kind == StackFrameKind.CloseBracketToken);
+            Debug.Assert(commaTokens.All(static t => t.Kind == StackFrameKind.CommaToken));
+
+            OpenBracket = openBracket;
+            CloseBracket = closeBracket;
+            CommaTokens = commaTokens;
+        }
+
+        internal override int ChildCount => 2 + CommaTokens.Length;
+
+        public override void Accept(IStackFrameNodeVisitor visitor)
+            => visitor.Visit(this);
+
+        internal override StackFrameNodeOrToken ChildAt(int index)
+        {
+            if (index == 0)
+            {
+                return OpenBracket;
+            }
+
+            if (index == ChildCount - 1)
+            {
+                return CloseBracket;
+            }
+
+            return CommaTokens[index - 1];
+        }
+    }
+
+    /// <summary>
+    /// The type argument list for a method declaration. 
+    /// 
+    /// <code>
+    /// Ex: MyType.MyMethod[T, U, V](T t, U u, V v) 
+    ///                    ^-----------------------  "[" = Open Token 
+    ///                     ^------^   ------------  "T, U, V" = SeparatedStackFrameNodeList&lt;StackFrameTypeArgumentNode&gt;
+    ///                             ^--------------  "]" = Close Token
+    /// </code>
+    /// 
+    /// </summary>
+    internal sealed class StackFrameTypeArgumentList : StackFrameNode
+    {
+        public readonly StackFrameToken OpenToken;
+        public readonly EmbeddedSeparatedSyntaxNodeList<StackFrameKind, StackFrameNode, StackFrameIdentifierNameNode> TypeArguments;
```

Comment:
have tests with weird generics.  like `<A, B<C>>`.  (just to make sure you don't crash).